diff --git a/util.c b/util.c
index 1328bed..71cccb5 100644
--- a/util.c
+++ b/util.c
@@ -2193,9 +2193,10 @@ static bool configure_stratum_mining(struct pool *pool)
 	snprintf(s, RBUFSIZE,
 		 "{\"id\": %d, \"method\": \"mining.configure\", \"params\": "
 		 "[[\""STRATUM_VERSION_ROLLING"\"], "
-		 "{\""STRATUM_VERSION_ROLLING".mask\": \"%x\""
+		 "{\""STRATUM_VERSION_ROLLING".mask\": \"%x\", "
+		 "\""STRATUM_VERSION_ROLLING".min-bit-count\": %d"
 		 "}]}",
-	  swork_id++, 0xffffffff);
+	  swork_id++, 0xffffffff, 2);
 
 	if (__stratum_send(pool, s, strlen(s)) != SEND_OK) {
 		applog(LOG_DEBUG, "Failed to send mining.configure");
-- 
2.7.4

